dali/operators/numba_function/numba_func.cc

```diff
@@ -0,0 +1,134 @@
+
+  types.void(
+    types.CPointer(OUTPUT_DTYPE), # Pointer to output sample
+    types.CPointer(types.int64),  # Pointer to output sample shape
+    types.CPointer(INPUT_DTYPE),  # Pointer to input sample
+    types.CPointer(types.int64),  # Pointer to input sample shape
+    types.int32                   # Number of sample dimensions
+  )
+
+)code", DALI_INT64)
+  .AddOptionalArg<int>("setup_fn", R"code(Pointer to function which should return output shape.
+
+If the setup function isn't provided then it's assumed output shape and type is the same as input.
+  
+The function should be a Numba C callback function (annotated with cfunc) with the following function signature::
+
+  types.void(
+    types.CPointer(types.int64), # Output batch shape pointer
+    types.int32,                 # Number of dimensions in output shape
+    types.CPointer(types.int32), # Output dtype pointer
+    types.CPointer(types.int64), # Input batch shape pointer
+    types.int32,                 # Number of dimensions in input shape
+    types.int32,                 # Input dtype
+    types.int32,                 # Number of samples in the batch
+    types.int32,                 # Number of outputs
+    types.int32                  # Number of inputs
+  )
+
+)code", 0);
+
+template <typename Backend>
+NumbaFunc<Backend>::NumbaFunc(const OpSpec &spec) : Base(spec) {
+  fn_ptr_ = spec.GetArgument<uint64_t>("fn_ptr");
+  setup_fn_ = spec.GetArgument<uint64_t>("setup_fn");
+}
+
+template<>
+void NumbaFunc<CPUBackend>::RunUserSetupFunc(std::vector<OutputDesc> &output_desc,
+    const workspace_t<CPUBackend> &ws) {
+  const auto &in = ws.InputRef<CPUBackend>(0);
+  if (!setup_fn_) {
+    output_desc[0] = {in.shape(), in.type()};
+    return;
+  }
+
+  auto in_shape = in.shape();
+  auto N = in_shape.num_samples();
+  auto ndim = in_shape.sample_dim();
+  TensorListShape<> output_shape;
+  output_shape.resize(N, ndim);
+  DALIDataType out_type = DALIDataType::DALI_NO_TYPE;
+  DALIDataType in_type = in.type().id();
+  ((void (*)(void*, int32_t, void*, const void*, int32_t, int32_t, int32_t))setup_fn_)(
+      output_shape.tensor_shape_span(0).data(), ndim, &out_type,
+      in_shape.tensor_shape_span(0).data(), ndim, in_type, N);
+
+  DALI_ENFORCE(out_type != DALIDataType::DALI_NO_TYPE,
+    "Output type was not set by the custom setup function.");
+  for (int i = 0; i < N; i++) {
+    for (int d = 0; d < ndim; d++) {
+      DALI_ENFORCE(output_shape.tensor_shape_span(i)[d] >= 0,
+        make_string(d, "-th", " dimension of ", i, "-th sample's shape is negative."));
+    }
+  }
+  output_desc[0].type = dali::TypeTable::GetTypeInfo(out_type);
+  output_desc[0].shape = output_shape;
+}
+
+template <>
+bool NumbaFunc<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+    const workspace_t<CPUBackend> &ws) {
+  const auto &in = ws.InputRef<CPUBackend>(0);
+  output_desc.resize(1);
+  RunUserSetupFunc(output_desc, ws);
```

Comment:
Why you need a separate `RunUserSetupFunc`, can we include its body just here?
This additional level of indirection doesn't seem to provide much of the value.